### PR TITLE
add reject from origin domain feature

### PR DESCRIPTION
**Pull request by cyd01**
Add a new feature to be able to reject email from specific domains.  
The list of domains to reject is defined in a new environment variable `INBUCKET_SMTP_REJECTORIGINDOMAINS`.
This variable is a coma separated strings list (like `INBUCKET_SMTP_ACCEPTDOMAINS`, `INBUCKET_SMTP_REJECTDOMAINS`, ...).

Signed-off-by: Cyril Dupont <cyd@9bis.com>

### CODE REVIEW

File: pkg/server/smtp/handler.go [hunk 1]
```diff
@@ -384,7 +385,16 @@ func (s *Session) readyHandler(cmd string, arg string) {
 			return
 		}
 		from := m[1]
+		s.logger.Debug().Msgf("Mail sender is %v", from)
 		localpart, domain, err := policy.ParseEmailAddress(from)
+		s.logger.Debug().Msgf("Domain sender is %v", domain)
+
+		if stringutil.SliceContains(s.Server.config.RejectOriginDomains, domain) {
```
Comment (jhillyerd) [inline pkg/server/smtp/handler.go:392]: This should use the new method you defined above in address.go.  In the future we could add a `sender.go` or `origin.go` policy file similar to the existing `recipient.go` to encapsulate any checks we want to do, but you don't have to do this unless you want to.

**Review by jhillyerd — CHANGES_REQUESTED** [review]
This needs test coverage.  I'd suggest cloning the `TestReadyStateInvalidCommands` func in handler_test.go and modifying it to do so.

File: pkg/server/smtp/handler.go (hunk 1 above)
Comment (cyd01) [reply pkg/server/smtp/handler.go]: Ok  
I've choosen to create this new origin.o file.  
I could not succeed to make requested changes in address.go.

File: pkg/policy/origin.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+// Origin represents a potential email origin, allows policies for it to be queried.
+type Origin struct {
+	mail.Address
+	addrPolicy *Addressing
+	// LocalPart is the part of the address before @, including +extension.
+	LocalPart string
+	// Domain is the part of the address after @.
+	Domain string
+	// Mailbox is the canonical mailbox name for this recipient.
+	Mailbox string
```
Comment (jhillyerd) [inline pkg/policy/origin.go:16]: I don't think Mailbox will needed for Origin, it's referring to the Inbucket mailbox name.

**Review by jhillyerd — CHANGES_REQUESTED** [review]
Looks good overall.  But the tests no longer compile.  If you are on Linux or similar, you can run the entire suite by just typing `make`

**Comment by jhillyerd** [comment]
Let me know if there is anything I can do to help with this one, seems like we are pretty close.

**Comment by cyd01** [comment]
I'm sorry for not answering. I'm on summer hollydays. I'll be back at work next week.

File: pkg/policy/address.go [hunk 3]
```diff
@@ -71,6 +71,26 @@ func (a *Addressing) NewRecipient(address string) (*Recipient, error) {
 	}, nil
 }
 
+// SetOrigin parses an address into a Origin. This is used for parsing MAIL FROM argument,
+// not To headers.
+func (a *Addressing) SetOrigin(address string) (*Origin, error) {
```
[the remark this answers is not in the document]
Comment (cyd01) [reply pkg/policy/address.go]: Done

File: pkg/policy/origin.go (hunk 2 above)
Comment (cyd01) [reply pkg/policy/origin.go]: Done

**Comment by cyd01** [comment]
I've tryed to fix the issue with tests compilation (in file `pkg/message/manager_test.go`).  
But now the Github build process does not work (it work on my side with a simple `go build ./cmd/inbucket/` command).  
it seems I need help.

**Comment by jhillyerd** [comment]
It looks like you did fix the compilation.  Go doesn't run tests when you compile, but you can run the smtp tests with

`go test ./pkg/server/smtp`

looks like we are getting an unexpected SMTP response code

```
--- FAIL: TestEmptyEnvelope (0.00s)
    writer.go:25: {"level":"info","module":"smtp","remote":"pipe","session":17,"message":"Starting SMTP session"}
    writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":17,"message":"Line received: HELO localhost"}
    writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":17,"message":"Entering state READY"}
    writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":17,"message":"Line received: MAIL FROM:<>"}
    writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":17,"message":"Mail sender is "}
    writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":17,"message":"Origin domain is "}
    writer.go:25: {"level":"warn","module":"smtp","remote":"pipe","session":17,"from":"unspecified","error":"Domain part validation failed","message":"Bad address as MAIL arg"}
    writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":17,"message":"Line received: QUIT"}
    handler_test.go:98: Step 1, sent "MAIL FROM:<>", expected 250, got 501: "Bad origin address syntax"
    writer.go:25: {"level":"info","module":"smtp","remote":"pipe","session":18,"message":"Starting SMTP session"}
    writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":18,"message":"Line received: HELO localhost"}
    writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":18,"message":"Entering state READY"}
    writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":18,"message":"Line received: MAIL FROM: <>"}
    writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":18,"message":"Mail sender is "}
    writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":18,"message":"Origin domain is "}
    writer.go:25: {"level":"warn","module":"smtp","remote":"pipe","session":18,"from":"unspecified","error":"Domain part validation failed","message":"Bad address as MAIL arg"}
    writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":18,"message":"Line received: QUIT"}
    handler_test.go:107: Step 1, sent "MAIL FROM: <>", expected 250, got 501: "Bad origin address syntax"
    writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":18,"message":"Entering state QUIT"}
    writer.go:25: {"level":"info","module":"smtp","remote":"pipe","session":18,"message":"Closing connection"}
    writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":17,"message":"Entering state QUIT"}
    writer.go:25: {"level":"info","module":"smtp","remote":"pipe","session":17,"message":"Closing connection"}
--- FAIL: TestReadyStateRejectedDomains (0.00s)
    --- FAIL: TestReadyStateRejectedDomains/MAIL_FROM_john@validdomain.com (0.00s)
        writer.go:25: {"level":"info","module":"smtp","remote":"pipe","session":35,"message":"Starting SMTP session"}
        writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":35,"message":"Line received: HELO localhost"}
        writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":35,"message":"Entering state READY"}
        writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":35,"message":"Line received: MAIL FROM john@validdomain.com"}
        writer.go:25: {"level":"warn","module":"smtp","remote":"pipe","session":35,"message":"Bad MAIL argument: \"FROM john@validdomain.com\""}
        writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":35,"message":"Line received: QUIT"}
        handler_test.go:219: Step 1, sent "MAIL FROM john@validdomain.com", expected 250, got 501: "Was expecting MAIL arg syntax of FROM:<address>"
        writer.go:25: {"level":"debug","module":"smtp","remote":"pipe","session":35,"message":"Entering state QUIT"}
        writer.go:25: {"level":"info","module":"smtp","remote":"pipe","session":35,"message":"Closing connection"}
FAIL
FAIL    github.com/inbucket/inbucket/pkg/server/smtp    0.011s
FAIL
```

**Review by jhillyerd — CHANGES_REQUESTED** [review]
Looks great overall, just a couple small tweaks

File: pkg/policy/address.go [hunk 4]
```diff
@@ -99,6 +117,15 @@ func (a *Addressing) ShouldStoreDomain(domain string) bool {
 	return false
 }
 
+// ShouldAcceptOriginDomain indicates if Inbucket accept mail from the specified domain.
+func (a *Addressing) ShouldAcceptOriginDomain(domain string) bool {
+	domain = strings.ToLower(domain)
+	if stringutil.SliceContains(a.Config.SMTP.RejectOriginDomains, domain) {
```
[the remark this answers is not in the document]
Comment (cyd01) [reply pkg/policy/address.go:123]: Much better indeed

**Review by jhillyerd — APPROVED** [review]

**Comment by jhillyerd** [comment]
Thank you!

**Pull request merged**